proposals/NNNN-task-priority-escalation-apis.md

```diff
@@ -0,0 +1,197 @@
+    var task: Task<Void, any Error>? // unstructured upstream consumer task
+    var buffer: Deque<Element>
+    var upstreamContinuations: [UnsafeContinuation<Void, Error>]
+    var downstreamContinuation: UnsafeContinuation<Element?, Error>?
+  }
+  
+  let state = Mutex<State>(State())
+  
+  func next() async throws {
+    self.state.withLock { state in
+      if state.task == nil {
+        state.task = Task {
+          // Consume from the base iterators
+          // ... 
+        }
+      }
+    }
+
+    if let element = self.state.withLock { $0.buffer.popFirst() } {
+      return element
+    } else {
+      // We are handling cancellation here and need to handle task escalation here as well
+      try await withTaskCancellationHandler { 
+        // HERE: need to handle priority escalation and boost `state.task`
+        try await withCheckedContinuation { cont in
+          self.state.withLock { $0.consumerContinuation = cont }
+        }
+      } onCancel: {
+        // trigger cancellation of tasks and fail continuations
+      }
+    }
+  }
+}
+```
+
+The above example showcases a common pattern: often an unchecked continuation is paired with a Task used to complete it. Around the suspension on the continuation, waiting for it to be resumed, developers often install a task cancellation handler in order to potentially break out of potentially unbounded waiting for a continuation to be resumed. Around the same suspension (marked with `HERE` in the snippet above), we might want to insert a task priority escalation handler in order to priority boost the task that is used to resume the continuation. This can be important for correctness and performance of such operations, so we should find a way to offer these libraries to participate in task priority handling.
+
+Another example of libraries which may want to reach for manual task priority escalation APIs are libraries which facilitate communication across process boundaries, and would like to react to priority escalation and propagate it to a different process. Relying on the built-in priority escalation mechanisms won't work, because they are necessarily in-process, so libraries like this need to be able to participate and be notified when priority escalation happens, and also be able to efficiently cause the escalation inside the other process.
+
+## Proposed solution
+
+In order to address the above use-cases, we propose to add a pair of APIs: to react to priority escalation happening within a block of code, and an API to _cause_ a priority escalation without resorting to trickery using creating new tasks whose only purpose is to escalate the priority of some other task:
+
+```swift
+let m: Mutex<Task<Void, Never>?> 
+
+await withTaskPriorityEscalationHandler {
+  await withCheckedContinuation { cc in 
+    let t = Task { cc.resume() }
+    m.withLock { $0 = t }
+  } onPriorityEscalated: { newPriority in 
+    let t = m.withLock { $0 }
+    Task.escalatePriority(t, to: newPriority)
```

Comment:
I think you've got a type mismatch in this example, `t` is an `Optional`, given the type of `m`.